### ISSUE TITLE
ia básica 

### ISSUE THREAD

**Issue by paulohenriquecostafonseca61-sudo** [paulohenriquecostafonseca61-sudo/IA-B-SICA-#1]
import random
import streamlit as st

st.set_page_config(page_title="Como pensa uma IA simples", page_icon="🧠", layout="centered")

# ---------- ESTILO ----------
st.markdown("""
<style>
    .stApp { background-color: #1B2B22; color: #F2EFE6; }
    h1, h2, h3 { font-family: Georgia, serif; }
    .kicker { color: #7FA98F; font-family: monospace; font-size: 0.85rem; }
    .dim { color: #C7CFC4; }
    .part-mark { color: #E8B33D; font-style: italic; font-size: 1.4rem; }
</style>
""", unsafe_allow_html=True)

# ---------- HERO ----------
st.markdown('<div class="kicker">um guia curto e prático</div>', unsafe_allow_html=True)
st.title("Como pensa uma IA simples")
st.markdown(
    '<p class="dim">Por baixo de qualquer sistema complexo de inteligência artificial existe uma ideia '
    'bem pequena: um número que decide, com base em outros números, se algo é "sim" ou "não". '
    'Este guia explica essa ideia — e mostra como programar a sua primeira versão dela.</p>',
    unsafe_allow_html=True,
)

st.divider()

# ---------- CONCEITO ----------
st.header("A unidade mais simples: o perceptron")
st.write(
    "Em 1958, o psicólogo Frank Rosenblatt criou um modelo matemático inspirado num neurônio, "
    "chamado **perceptron**. Ele continua sendo, hoje, o jeito mais direto de entender qualquer "
    "rede neural — mesmo as gigantes por trás dos assistentes de IA atuais."
)
st.markdown(
    '<p class="dim">A ideia: cada entrada (x1, x2, x3...) é multiplicada por um peso, os resultados '
    'são somados, e se a soma passar de um certo limiar, a saída é 1. Caso contrário, é 0. '
    'É basicamente uma votação ponderada.</p>',
    unsafe_allow_html=True,
)

st.divider()

# ---------- QUATRO PEÇAS ----------
st.header("As quatro peças de qualquer IA")
st.markdown('<p class="dim">Da IA mais simples até os modelos mais avançados, as mesmas quatro peças aparecem — só a escala muda.</p>', unsafe_allow_html=True)

pecas = [
    ("01", "Dados", "Exemplos que mostram à IA como o problema se comporta. Para prever se chove, seriam medições de umidade, pressão e temperatura de dias passados."),
    ("02", "Pesos", "Números que dizem o quanto cada entrada importa. No começo são aleatórios; o treinamento existe justamente para ajustá-los."),
    ("03", "Função de ativação", "Decide como transformar a soma ponderada numa resposta — um \"sim/não\", uma probabilidade, ou uma palavra, dependendo do modelo."),
    ("04", "Treinamento", "O processo de comparar a resposta da IA com a resposta certa, e ajustar os pesos um pouco a cada erro, repetidas vezes."),
]
for num, titulo, texto in pecas:
    c1, c2 = st.columns([1, 8])
    c1.markdown(f'<div class="part-mark">{num}</div>', unsafe_allow_html=True)
    c2.markdown(f"**{titulo}**")
    c2.markdown(f'<span class="dim">{texto}</span>', unsafe_allow_html=True)

st.divider()

# ---------- DEMO INTERATIVA (tudo em Python) ----------
st.header("Experimente um perceptron")
st.write("Mova os controles abaixo para simular duas entradas e seus pesos. O resultado mostra o que o perceptron decide — calculado ao vivo em Python.")

THRESHOLD = 0.3

col1, col2 = st.columns(2)
with col1:
    x1 = st.slider("entrada x1", 0.0, 1.0, 0.6, 0.05)
    x2 = st.slider("entrada x2", 0.0, 1.0, 0.4, 0.05)
with col2:
    w1 = st.slider("peso w1", -1.0, 1.0, 0.5, 0.05)
    w2 = st.slider("peso w2", -1.0, 1.0, -0.3, 0.05)

soma = x1 * w1 + x2 * w2
ativado = soma > THRESHOLD

st.metric("soma ponderada", f"{soma:.2f}")
if ativado:
    st.success("Saída: **1** (a soma ultrapassou o limiar de 0.3)")
else:
    st.info("Saída: **0** (a soma não ultrapassou o limiar de 0.3)")

st.divider()

# ---------- CÓDIGO PASSO A PASSO ----------
st.header("Programando o seu, do zero")
st.write("Sem nenhuma biblioteca de IA — só matemática básica em Python. Este perceptron aprende a distinguir dois grupos de pontos.")

st.subheader("1. Defina os pesos e o limiar")
st.code(
"""import random

pesos = [random.uniform(-1, 1) for _ in range(2)]
limiar = 0.0
taxa_aprendizado = 0.1""",
    language="python",
)

st.subheader("2. Escreva a função de previsão")
st.code(
"""def prever(entradas, pesos, limiar):
    soma = sum(e * p for e, p in zip(entradas, pesos))
    return 1 if soma > limiar else 0""",
    language="python",
)

st.subheader("3. Treine com exemplos")
st.code(
"""dados = [
    ([0.9, 0.8], 1),
    ([0.1, 0.2], 0),
    ([0.8, 0.1], 1),
    ([0.2, 0.9], 0),
]

for epoca in range(50):
    for entradas, esperado in dados:
        previsto = prever(entradas, pesos, limiar)
        erro = esperado - previsto
        for i in range(len(pesos)):
            pesos[i] += taxa_aprendizado * erro * entradas[i]
        limiar -= taxa_aprendizado * erro""",
    language="python",
)

st.subheader("4. Teste o resultado")
st.code(
"""print(prever([0.85, 0.75], pesos, limiar))  # provavelmente 1
print(prever([0.15, 0.25], pesos, limiar))  # provavelmente 0""",
    language="python",
)

# ---------- TREINAMENTO AO VIVO ----------
st.subheader("Bônus: rode o treinamento agora mesmo")
if st.button("Treinar perceptron"):
    pesos = [random.uniform(-1, 1) for _ in range(2)]
    limiar = 0.0
    taxa_aprendizado = 0.1
    dados = [
        ([0.9, 0.8], 1),
        ([0.1, 0.2], 0),
        ([0.8, 0.1], 1),
        ([0.2, 0.9], 0),
    ]

    def prever(entradas, pesos, limiar):
        s = sum(e * p for e, p in zip(entradas, pesos))
        return 1 if s > limiar else 0

    for _ in range(50):
        for entradas, esperado in dados:
            previsto = prever(entradas, pesos, limiar)
            erro = esperado - previsto
            for i in range(len(pesos)):
                pesos[i] += taxa_aprendizado * erro * entradas[i]
            limiar -= taxa_aprendizado * erro

    st.write(f"Pesos finais: `{[round(p, 3) for p in pesos]}` | Limiar final: `{round(limiar, 3)}`")
    st.write(f"Teste com [0.85, 0.75] → **{prever([0.85, 0.75], pesos, limiar)}** (esperado: 1)")
    st.write(f"Teste com [0.15, 0.25] → **{prever([0.15, 0.25], pesos, limiar)}** (esperado: 0)")

st.divider()

# ---------- PRÓXIMOS PASSOS ----------
st.header("Pra ir além de um único neurônio")
st.markdown('<p class="dim">Um perceptron só resolve problemas simples, separáveis por uma linha reta. Modelos reais empilham milhares deles em camadas. Os próximos passos naturais:</p>', unsafe_allow_html=True)

caminhos = [
    ("numpy", "Para fazer as mesmas contas com vetores e matrizes, muito mais rápido do que loops em Python puro."),
    ("scikit-learn", "Biblioteca com modelos prontos (árvores de decisão, regressão, redes simples) para praticar sem reescrever a matemática toda vez."),
    ("PyTorch", "Framework usado para redes neurais de verdade, com camadas empilhadas e treinamento automático dos pesos."),
    ("Um dataset real", "Experimente treinar algo com dados de verdade — o MNIST (dígitos escritos à mão) é o ponto de partida clássico."),
]
for titulo, texto in caminhos:
    st.markdown(f"**{titulo}**")
    st.markdown(f'<span class="dim">{texto}</span>', unsafe_allow_html=True)

st.divider()
st.caption("Feito para explicar a ideia por trás de qualquer IA — do perceptron de 1958 até os modelos de hoje.")

**Comment by paulohenriquecostafonseca61-sudo** [comment]
Me perdoem se estiver ridículo pois é o meu primeiro projeto
